### PR TITLE
[stdlib] Added Armv5 support

### PR DESCRIPTION
**Pull request by colemancda**
Added support for cross compiling the Swift Stdlib for Armv5 processors.

Resolves [SR-15944](https://bugs.swift.org/browse/SR-15944).

### CODE REVIEW

**Comment by colemancda** [comment]
The only issues I have encountered is `/stdlib/public/Platform/4/Glibc.swift:14:19: error: no such module 'SwiftGlibc'`. I've seen previous threads about older cmake versions and Android not being able to find the headers. Not sure why I am having this issue.

```
coleman@macbook-linux:~/Developer/buildroot$ /home/coleman/Developer/swift-install/usr/bin/swiftc -emit-module -o /home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/./lib/swift/linux/Glibc.swiftmodule/armv5te-unknown-linux-gnueabi.swiftmodule -avoid-emit-module-source-info -sdk /home/coleman/Developer/buildroot/output/host/arm-buildroot-linux-gnueabi/sysroot -target armv5te-unknown-linux-gnueabi -resource-dir /home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/./lib/swift -O -D SWIFT_ENABLE_EXPERIMENTAL_CONCURRENCY -D SWIFT_RUNTIME_OS_VERSIONING -module-cache-path /home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/./module-cache -no-link-objc-runtime -enable-library-evolution -Xfrontend -enforce-exclusivity=unchecked -module-name Glibc -swift-version 5 -swift-version 5 -autolink-force-load -runtime-compatibility-version none -disable-autolinking-runtime-compatibility-dynamic-replacements -Xfrontend -disable-autolinking-runtime-compatibility-concurrency -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -warn-implicit-overrides -Xfrontend -disable-implicit-concurrency-module-import -module-link-name swiftGlibc -whole-module-optimization -parse-as-library -I /home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/./lib/swift/linux -emit-module-interface-path /home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/./lib/swift/linux/Glibc.swiftmodule/armv5te-unknown-linux-gnueabi.swiftinterface -Xfrontend -experimental-skip-non-inlinable-function-bodies @/home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/stdlib/public/Platform/10ae097bc634fb611f4252af75686b2b9426d4a8.txt
/home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/stdlib/public/Platform/4/Glibc.swift:14:19: error: no such module 'SwiftGlibc'
@_exported import SwiftGlibc // Clang module
```

Building with trunk buildroot and Swift.
- Buildroot config 
https://gist.github.com/colemancda/6cf77d988b5207c35c54e595c69ad776
- Buildroot patches to add Swift
https://gist.github.com/colemancda/e2d7923b51697edec61c0004dfece150
- Swift 5.5.3 armv5 patches (automatically applied by Buildroot)
https://gist.github.com/colemancda/ded1de5b1b84a5b84a13c2433b9001a7
- Dispatch 5.5.3 armv5 patches

[swift-armv5-build-error.txt](https://github.com/apple/swift/files/8190560/swift-armv5-build-error.txt)

**Comment by colemancda** [comment]
@buttaface Any suggestions? This seems to be a cmake error, I checked and both 
`'/home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/lib/swift/linux/armv5/SwiftGlibc.h'
`'/home/coleman/Developer/buildroot/output/build/swift-5.5.3/build/lib/swift/linux/armv5/glibc.modulemap'`
Exists and have the expected contents.
[build-files.zip](https://github.com/apple/swift/files/8190567/build-files.zip)

**Comment by buttaface** [comment]
Are you recompiling the Swift compiler with this pull? That's the only thing I can think of, that the unpatched compiler doesn't know this new arch.

**Comment by colemancda** [comment]
Yes, I am recompiling the compiler, mainly since [lib/IDE/Utils.cpp](https://github.com/apple/swift/pull/41686#diff-e95d9a20bd94dd81924ef45bd4c87c1f9cce1e619c721c1d977dfe2619842e0c) and [lib/Basic/Platform.cpp](https://github.com/apple/swift/pull/41686#diff-52d2619d68342be1dd3736cc6d7b7da3b2badbbb2c7cc133a8d72cf41a2ef02b) are part of the compiler.

**Comment by buttaface** [comment]
I compared your failing command to a local trunk build on my Android AArch64 phone and I notice you're passing a target of `-target armv5te-unknown-linux-gnueabi` but your `glibc.modulemap` is in `lib/swift/linux/armv5/`. Are you sure the Swift compiler knows that the two arch names are used interchangeably?

Try running the failing command manually, but add these two flags `-Xfrontend -dump-clang-diagnostics`. That will tell you what path it's checking for the `glibc.modulemap`.

File: lib/Basic/Platform.cpp [hunk 1]
```diff
@@ -218,6 +218,8 @@ StringRef swift::getMajorArchitectureName(const llvm::Triple &Triple) {
       return "armv7";
     case llvm::Triple::SubArchType::ARMSubArch_v6:
       return "armv6";
+    case llvm::Triple::SubArchType::ARMSubArch_v5:
```
Comment (colemancda) [inline lib/Basic/Platform.cpp:221]: This is similar to https://github.com/apple/swift-driver/blob/0f9f247c8cc3d225dbbc263b1ec082bf2c8675fe/Sources/SwiftDriver/Jobs/GenericUnixToolchain%2BLinkerSupport.swift#L45 and its possible Swift-Driver needs to be patched to find Glibc.

Comment (colemancda) [reply lib/Basic/Platform.cpp:221]: Patched 
https://gist.github.com/colemancda/df87bab835f431f6b02ea9c795dffcd7
https://github.com/apple/swift-driver/pull/1027

**Comment by colemancda** [comment]
[Error building with Buildroot](https://github.com/apple/swift/files/8192009/swift-armv5-buildroot-error.txt)

**Comment by colemancda** [comment]
@buttaface I was able to get the Swift stdlib compiling successfully by changing the armv5te triple  to armv5. Thanks!

**Comment by colemancda** [comment]
@CodaFi I have been able to successfully cross compile the Stdlib for armv5 (including libdispatch and Concurrency module). Can we have the CI build this?

**Comment by colemancda** [comment]
SwiftPM patches
https://github.com/apple/swift-tools-support-core/pull/296

**Comment by colemancda** [comment]
I am able to build a Hello world program in Swift for Armv5

```
coleman@macbook-linux:~/Developer/buildroot$ readelf -A /home/coleman/Developer/buildroot/output/target/usr/bin/hello
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "5TE"
  Tag_CPU_arch: v5TE
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-1
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_enum_size: int
```

**Comment by CodaFi** [comment]
⛵

**Pull request merged**